File: Javascript/merge_sort.js
const mergeSort = function(array) {
  if (array.length < 2) {
      return array;
    } else {
      const middle = Math.floor(array.length / 2);
      const left = mergeSort(array.slice(0, middle));
      const right = mergeSort(array.slice(middle));

      return merge(left, right);
    }
}

const merge = function(left, right){
  const merged = []
  while(left.length > 0 && right.length > 0){
    let nextItem = (left[0] < right[0]) ? left.shift() : right.shift();
    merged.push(nextItem)
  }

  return merged.concat(left, right)
}

let array = [5,4,3,2,1]
console.log(mergeSort(array))

File: Ruby/merge_sort.rb
def merge_sort(arr, &prc)
  return arr if arr.length <= 1
  prc ||= Proc.new {|x, y| x < y}
  mid = arr.length / 2
  left = merge_sort(arr[0, mid], &prc)
  right = merge_sort(arr[mid, arr.length], &prc)

  merge!(left, right, &prc)
end

def merge!(left, right, &prc)
  result = []
  until left.empty? || right.empty?
    prc.call(left.first, right.first) ? result << left.shift : result << right.shift
  end

  result + left + right
end
